File: monolith-spigot/src/main/java/gg/warcraft/monolith/spigot/core/SpigotTaskService.java
package gg.warcraft.monolith.spigot.core;

import com.google.inject.Inject;
import gg.warcraft.monolith.api.core.Task;
import gg.warcraft.monolith.api.core.TaskService;
import gg.warcraft.monolith.api.core.Duration;
import gg.warcraft.monolith.api.util.TimeUtils;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

public class SpigotTaskService implements TaskService {
    private final Plugin plugin;

    @Inject
    public SpigotTaskService(Plugin plugin) {
        this.plugin = plugin;
    }

    private BukkitRunnable wrapBukkitRunnable(Runnable runnable) {
        return new BukkitRunnable() {
            @Override
            public void run() {
                runnable.run();
            }
        };
    }

    @Override
    public Task runLater(Runnable runnable, Duration delay) {
        BukkitRunnable bukkitRunnable = wrapBukkitRunnable(runnable);
        BukkitTask bukkitTask = bukkitRunnable.runTaskLater(plugin, delay.inTicks());
        return new SpigotTask(bukkitTask);
    }

    @Override
    public Task runLaterAsync(Runnable runnable, Duration delay) {
        BukkitRunnable bukkitRunnable = wrapBukkitRunnable(runnable);
        BukkitTask bukkitTask = bukkitRunnable.runTaskLaterAsynchronously(plugin, delay.inTicks());
        return new SpigotTask(bukkitTask);
    }

    @Override
    public Task runNextTick(Runnable runnable) {
        BukkitRunnable bukkitRunnable = wrapBukkitRunnable(runnable);
        BukkitTask bukkitTask = bukkitRunnable.runTask(plugin);
        return new SpigotTask(bukkitTask);
    }

    @Override
    public Task runNextTickAsync(Runnable runnable) {
        BukkitRunnable bukkitRunnable = wrapBukkitRunnable(runnable);
        BukkitTask bukkitTask = bukkitRunnable.runTaskAsynchronously(plugin);
        return new SpigotTask(bukkitTask);
    }

    @Override
    public Task runTask(Runnable runnable, Duration delay, Duration period) {
        BukkitRunnable bukkitRunnable = wrapBukkitRunnable(runnable);
        BukkitTask bukkitTask = bukkitRunnable.runTaskTimer(plugin, delay.inTicks(), period.inTicks());
        return new SpigotTask(bukkitTask);
    }

    @Override
    public Task runTaskAsync(Runnable runnable, Duration delay, Duration period) {
        BukkitRunnable bukkitRunnable = wrapBukkitRunnable(runnable);
        BukkitTask bukkitTask = bukkitRunnable.runTaskTimerAsynchronously(plugin, delay.inTicks(), period.inTicks());
        return new SpigotTask(bukkitTask);
    }

    @Override
    public Task runEachTick(Runnable runnable) {
        return runTask(runnable, Duration.oneTick(), Duration.oneTick());
    }

    @Override
    public Task runEachTickAsync(Runnable runnable) {
        return runTaskAsync(runnable, Duration.oneTick(), Duration.oneTick());
    }
}

File: monolith-spigot/src/main/scala/gg/warcraft/monolith/spigot/core/task/SpigotTaskService.scala
/*
 * MIT License
 *
 * Copyright (c) 2020 WarCraft <https://github.com/WarCraft>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package gg.warcraft.monolith.spigot.core.task

import gg.warcraft.monolith.api.core.task.{Task, TaskService}
import gg.warcraft.monolith.api.core.Duration
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.{BukkitRunnable, BukkitTask}

class SpigotTaskService(implicit plugin: Plugin) extends TaskService {
  private def schedule(
      task: () => Unit,
      scheduler: BukkitRunnable => BukkitTask
  ): Task = {
    val bukkitTask = scheduler(() => task())
    new SpigotTask(bukkitTask)
  }

  override def evalLater(delay: Duration, task: => Unit): Task =
    schedule(() => task, _.runTaskLater(plugin, delay.ticks))

  override def evalAsync(task: => Unit): Task =
    schedule(() => task, _.runTaskAsynchronously(plugin))

  override def runTask(period: Duration, task: () => Unit): Task =
    schedule(task, _.runTaskTimer(plugin, 0, period.ticks))

  override def runTaskAsync(period: Duration, task: () => Unit): Task =
    schedule(task, _.runTaskTimerAsynchronously(plugin, 0, period.ticks))
}
